qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+    """
+    def visit(self, node: Any):
+        """Visit a node."""
+        visitor = self._get_visitor(node.__class__)
+        return visitor(node)
+
+    def _get_visitor(self, node_class):
+        """A helper function to recursively investigate superclass visitor method."""
+        if node_class == object:
+            return self.generic_visit
+
+        try:
+            return getattr(self, f'visit_{node_class.__name__}')
+        except AttributeError:
+            # check super class
+            return self._get_visitor(node_class.__base__)
+
+    def visit_ScheduleBlock(self, node: ScheduleBlock):
+        """Visit ``ScheduleBlock``. Recursively visit context blocks and overwrite.
+
+        .. note:: ``ScheduleBlock`` can have parameters in blocks and its alignment.
+        """
+        raise NotImplementedError
+
+    def visit_Schedule(self, node: Schedule):
+        """Visit ``Schedule``. Recursively visit schedule children and overwrite."""
+        raise NotImplementedError
+
+    def generic_visit(self, node: Any):
+        """Called if no explicit visitor function exists for a node."""
+        raise NotImplementedError
+
+
+class ParameterSetter(NodeVisitor):
+    """Node visitor for parameter binding.
+
+    This visitor is initialized with a dictionary of parameters to be assigned,
+    and assign values to operands of nodes found.
+    """
+    def __init__(self, param_map: Dict[ParameterExpression, ParameterValueType]):
+        self._param_map = param_map
+
+    # Top layer: Assign parameters to programs
+
+    def visit_ScheduleBlock(self, node: ScheduleBlock):
+        """Visit ``ScheduleBlock``. Recursively visit context blocks and overwrite.
+
+        .. note:: ``ScheduleBlock`` can have parameters in blocks and its alignment.
+        """
+        # accessing to protected member
+        node._blocks = [self.visit(block) for block in node.instructions]
+        node._alignment_context = self.visit(node.alignment_context)
+
+        self._update_parameter_manager(node)
+        return node
+
+    def visit_Schedule(self, node: Schedule):
+        """Visit ``Schedule``. Recursively visit schedule children and overwrite."""
+        # accessing to private member
+        node._Schedule__children = [(t0, self.visit(sched)) for t0, sched in node.instructions]
+
+        # update timeslots
+        for chan in copy(node.channels):
+            if isinstance(chan.index, ParameterExpression):
+                chan_timeslots = node._timeslots.pop(chan)
+                new_channel = self.visit_Channel(chan)
+
+                # Merge with existing channel
+                if new_channel in node.timeslots:
+                    sched = Schedule()
+                    sched._timeslots = {new_channel: chan_timeslots}
+                    node._add_timeslots(0, sched)
+                # Or add back under the new name
+                else:
+                    node._timeslots[new_channel] = chan_timeslots
```

Comment:
this is a little sketchy because it shouldn't rely on the private internal representations of another class. can this be moved into a dedicated method in `Schedule` which could be called here?

Reply:
That's reasonable suggestion. Let me try :)